testing-styleguide/README.md

```diff
@@ -0,0 +1,496 @@
+let(:hiring_updater) { instance_double ::Services::Hiring::Update::HiringUpdater }
+let(:job_updater) { instance_double ::Services::Job::Update::JobUpdater }
+
+# Use FactoryBot.build_stubbed to generate your data
+let(:hiring) { build_stubbed(:hiring) }
+
+context 'when manager is called valid params' do
+  let(:params) do
+    {
+      id: hiring.id,
+      status: Hiring.status.approved
+    }
+  end
+  let(:updated_hiring) do
+    hiring.update(status: Hiring.status.approved)
+    hiring
+  end
+
+  before do
+    # Replace HiringUpdater and JobUpdater in the subject under test
+    allow(subject).to receive(:hiring_updater).and_return hiring_updater
+    allow(subject).to receive(:job_updater).and_return job_updater
+
+    # Only allow calls to update that happen with the expected set of parameters
+    allow(hiring_updater).to receive(:update).with(params).and_return updated_hiring
+    allow(job_updater).to receive(:update).with(updated_hiring.job_id)
+
+    @result = subject.update(params)
+  end
+
+  it "should call the hiring updater with appropriate arguments the right number of times" do
+    # Assert that HiringUpdater.update has been called once
+    expect(hiring_updater).to have_received(:update).once
+  end
+
+  it "should call the job updater with appropriate arguments the right number of times" do
+    # Assert that JobUpdater.update has been called once
+    expect(job_updater).to have_received(:update).once
+  end
+
+  it "should return the updated Hiring" do
+    # Assert the return is the expected
+    expect(@result.status).to eq(Hiring.status.approved)
+  end
+
+  # ... remaining test scenarios
+end
+```
+
+A point about performance: sometimes we want to use Factories to faster, and more reliably, build our data. Whenever possible, use stubbed methods such as `FactoryBot.build_stubbed`. Methods like `FactoryBot.create` update the test database, which is very useful in integration tests but is also a very expensive operation; in scale, hitting the database in manager layer considerably slows down the test pipeline. If the manager is not supposed to directly access an ActiveRecord, avoid using create methods.
+
+#### 🛠️ Service layer
+
+Using the same logic as in the Manager layer, we mock the dependencies and test if they are called appropriately.
+
+Let's follow the example of the `HiringUpdater`. In order to update the ActiveRecord records, it needs to first validate the parameters received from the manager via a `HiringUpdateValidator`. Then it either raises an Exception if parameters are invalid or proceeds to call a `DelegateRepository` in order to effectively persist the updates to the database. After updating, it returns the updated Hiring.
+
+❌ Bad
+
+```ruby
+context 'when service is called with valid params' do
+  it 'updates the hiring' do
+    # ... code asserting if hiring is updated
+    # !! The repository that updates jobs already tests this
+  end
+end
+```
+
+✅ Good
+
+```ruby
```

Comment:
I disagree with this concept on principle because - to me - clearly while the cost of running the test is lower, the cost of understanding and changing it is way higher, and the risk of this giving us a false green is way higher too. A new developer will read this many times before getting what's going on.

Instead of overusing doubles this way, I would just avoid writing to the database by using `Factory.build(...)` instead of `Factory.create(...)`. It's a good compromise to me, and we can trust all the wires are connected correctly when we assert for behaviours.

By rewiring things with doubles and `receive(:call).with(kind_of(Something))` we are breaking the principles of **keeping the tests simple** and not re-implement the application logic.

I would change my mind after having a deep understanding of the codebase for `FactoryBot` as I have with Rails and Rspec. Since half of our codebase is (or it's gonna be) tests, we should use these tool with care. I encourage you to read the code for the [available factory strategies here](https://github.com/thoughtbot/factory_bot/tree/main/lib/factory_bot/strategy), and choose witch one would you rather debug.

Reply:
That's an interesting argument. Managers/Services tests (as proposed here) do rely on writing down, in the spec itself, what interactions are made with the following layers.

You mean instead of doing things like _allow(SomeService).to receive(:someMethod).with(someArgs).and_return aMockedResponse_ we could use the real implementation + a (lighter) FactoryBot strategy? 

Will check out the strategies implementations. I did a few tests in our codebase and it seems the test performance is usually driven by FactoryBot create vs build vs build_stubbed  - more than the doubling itself.

That may need some sort of refactoring in how we use FactoryBot though, as some nested factories are not using the parent strategy via association.

Reply:
@lucasmartins @joaophellip 

I believe that for this kind of Service we have two options here:

1. Testing without any mocks and use wisely the `FactoryBot`
2. Not testing at all. For me, this Service could be a simple method in the `Manager` layer, because it simpling calling two other classes that have their respective tests. Mocking and testing just the messages shared between them seems to me a bit redundant, not worth the time spent by the dev who wrote it, nor by a new person trying to understand it.